### DISCUSSION TITLE
How can I render a Quarto document as a Quarto Document (Qmd -> Qmd)?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#9842] [Q&A]
### Description

I'm interested in running a filter on the document that alters its own markdown **and** rendered output states via a filter. 

The [multi-format rendering](https://quarto.org/docs/output-formats/html-multi-format#overview) process indicates that if the format is found in the document's `format` key, then when using:

```sh
quarto render multi-format.qmd
```

We should expect all the formats to be listed in the `outdir`. 

Is there a possible way to get the modified `.qmd` file made available alongside the other formats? e.g.

```md
---
title: Sample Page
date: last-modified
toc: true
format: 
  html: default
  qmd: default # or identity? 
filter:
- alter
---
```

I experimented around with `md`; but, I'm not able to override the `.md` extension to `.qmd`

**Comment by cscheid** [comment]
> Is there a possible way to get the modified .qmd file made available alongside the other formats?

What do you mean by "the modified .qmd" file? Who's modifying it, and where?

**Reply by coatless** [reply]
> What do you mean by "the modified .qmd" file?

If we have a Qmd document and use a filter, then the document's AST should be changed. 

The modified Qmd document would contain these changes. 

> Who's modifying it?

Quarto extension would modify it. 

> And where? 

The modified `qmd` file would be placed inside the output directory that differs from the current Qmd location. 

----

I suppose I'm interested in knowing if there is an **identity-like transformation** where the Qmd document can be run through a filter and returned with the modified output as a Qmd file

**Reply by cscheid** [reply]
> I suppose I'm interested in knowing if there is an identity-like transformation where the Qmd document can be run through a filter and returned with the modified output as a Qmd file

No, there currently isn't; user filters are run after the engine executes, and in that case you're not going to get the .qmd file; you'll get a markdown file that (eg) could already have format-specific output. We have plans for a pre-engine filter step accessible to all documents, but it's not going to happen in the near future.

**Reply by coatless** [reply]
Two quick follow-ups:

1. If I wanted to take the risk on the `md` file having format-specific output, what would be the best way for an extension to change the output extension? (`.md` -> `.qmd`) 
2. Future is likely 1.6 or later?

**Reply by cscheid** [reply]
> If I wanted to take the risk on the md file having format-specific output, what would be the best way for an extension to change the output extension? (.md -> .qmd)

There's no built-in hook to do that, unfortunately. You could invoke `quarto render` multiple times manually to different formats, but Quarto itself doesn't have support for what you want to do.

**Reply by cscheid** [reply]
> Future is likely 1.6 or later?

No timeline to offer.

**Reply by mcanouil** [reply]
You could generate GFM as "qmd" leveraging `output-file` and/or `output-ext` but the file won't really be a Quarto document in the sense as it won't have Quarto specific syntax/features.

**Comment by d-morrison** [comment]
This would be helpful for me as well; I'm creating homework assignment files with solutions included, and then rendering them both with and without the solutions visible (using https://github.com/coatless-quarto/assign), and it would be great to be able to output the no-solutions version as a .qmd that I could give to students.
